diff --git a/odoo/modules/registry.py b/odoo/modules/registry.py
index 2473a80a..62a5813a 100644
--- a/odoo/modules/registry.py
+++ b/odoo/modules/registry.py
@@ -281,7 +281,10 @@ class Registry(Mapping):
         # on a fully loaded registry, and not on a registry being loaded.
         if self.ready:
             for model in env.values():
-                model._register_hook()
+                if model._name in ('multicompany.force.config', 'multicompany.force.security'):
+                    model._register_hook(update_module=True)
+                else:
+                    model._register_hook()
             env['base'].flush()
 
     @lazy_property
