diff --git a/odoo/addons/base/models/ir_translation.py b/odoo/addons/base/models/ir_translation.py
index c4d6915a..f149f1f2 100644
--- a/odoo/addons/base/models/ir_translation.py
+++ b/odoo/addons/base/models/ir_translation.py
@@ -44,6 +44,7 @@ class IrTranslationImport(object):
         # Note that Postgres will NOT inherit the constraints or indexes
         # of ir_translation, so this copy will be much faster.
         query = """ CREATE TEMP TABLE %s (
+                        id SERIAL PRIMARY KEY,
                         imd_model VARCHAR(64),
                         imd_name VARCHAR(128),
                         noupdate BOOLEAN
@@ -134,12 +135,17 @@ class IrTranslationImport(object):
                             WHERE EXCLUDED.value IS NOT NULL AND EXCLUDED.value != '';
                        """ % (self._model_table, self._table))
             count += cr.rowcount
-        cr.execute(""" INSERT INTO %s(name, lang, res_id, src, type, value, module, state, comments)
+        cr.execute(""" INSERT INTO {model}(name, lang, res_id, src, type, value, module, state, comments)
                        SELECT name, lang, res_id, src, type, value, module, state, comments
-                       FROM %s
-                       WHERE %%s OR noupdate is true
+                       FROM {table}
+                       WHERE ({where} OR noupdate is true) AND NOT EXISTS (
+                           SELECT FROM ONLY {model}
+                           WHERE type = {table}.type AND name = {table}.name AND lang = {table}.lang AND res_id = {table}.res_id
+                       )
                        ON CONFLICT DO NOTHING;
-                   """ % (self._model_table, self._table), [not self._overwrite])
+                   """.format(model=self._model_table,
+                              table=self._table,
+                              where='noupdate IS TRUE' if self._overwrite else 'TRUE'))
         count += cr.rowcount
 
         if self._debug:
