diff --git a/addons/bus/models/res_partner.py b/addons/bus/models/res_partner.py
index 71f2f8fb..3b7b9243 100644
--- a/addons/bus/models/res_partner.py
+++ b/addons/bus/models/res_partner.py
@@ -13,7 +13,7 @@ class ResPartner(models.Model):
     def _compute_im_status(self):
         self.env.cr.execute("""
             SELECT
-                U.partner_id as id,
+                IP.partner_id as id,
                 CASE WHEN max(B.last_poll) IS NULL THEN 'offline'
                     WHEN age(now() AT TIME ZONE 'UTC', max(B.last_poll)) > interval %s THEN 'offline'
                     WHEN age(now() AT TIME ZONE 'UTC', max(B.last_presence)) > interval %s THEN 'away'
@@ -21,8 +21,9 @@ class ResPartner(models.Model):
                 END as status
             FROM bus_presence B
             RIGHT JOIN res_users U ON B.user_id = U.id
-            WHERE U.partner_id IN %s AND U.active = 't'
-         GROUP BY U.partner_id
+            LEFT JOIN ir_property IP ON U.id = IP.user_id
+            WHERE IP.partner_id IN %s AND U.active = 't'
+         GROUP BY IP.partner_id
         """, ("%s seconds" % DISCONNECTION_TIMER, "%s seconds" % AWAY_TIMER, tuple(self.ids)))
         res = dict(((status['id'], status['status']) for status in self.env.cr.dictfetchall()))
         for partner in self:
