diff --git a/addons/hr_holidays/models/res_users.py b/addons/hr_holidays/models/res_users.py
index 24f70cde..a0fd5eda 100644
--- a/addons/hr_holidays/models/res_users.py
+++ b/addons/hr_holidays/models/res_users.py
@@ -52,12 +52,14 @@ class User(models.Model):
     @api.model
     def _get_on_leave_ids(self, partner=False):
         now = fields.Datetime.now()
-        field = 'partner_id' if partner else 'id'
-        self.env.cr.execute('''SELECT res_users.%s FROM res_users
+        field = 'ir_property.partner_id' if partner else 'res_users.id'
+        self.env.cr.execute('''SELECT %s FROM res_users
                             JOIN hr_leave ON hr_leave.user_id = res_users.id
+                            LEFT JOIN ir_property ON res_users.id = ir_property.user_id
+                            AND ir_property.company_id = %s
                             AND state not in ('cancel', 'refuse')
                             AND res_users.active = 't'
-                            AND date_from <= %%s AND date_to >= %%s''' % field, (now, now))
+                            AND date_from <= %%s AND date_to >= %%s''' % (field, self.env.company.id), (now, now))
         return [r[0] for r in self.env.cr.fetchall()]
 
     def _clean_leave_responsible_users(self):
